File: solutions/python/scale-generator/1/scale_generator.py
class Scale:
    
    def __init__(self, tonic):
        self.scale_sharp=['A', 'A#', 'B', 'C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#']
        self.scale_flat=['A', 'Bb', 'B', 'C', 'Db', 'D', 'Eb', 'E', 'F', 'Gb', 'G', 'Ab']
        #self.sharp_tonic=['G', 'D', 'A', 'E', 'B', 'F#', 'e', 'b', 'f#', 'c#']
        self.flat_tonic=['F', 'Bb', 'Eb', 'Ab', 'Db', 'Gb', 'd', 'g', 'c', 'f', 'bb', 'eb'] 
        self.tonic = tonic
        #print('tonic',tonic)

    def chromatic(self):
        scale = self.scale_sharp
        
        if self.tonic in self.flat_tonic:
            scale = self.scale_flat

        if self.tonic[0].islower():
            if len(self.tonic) == 1:
                si=scale.index(self.tonic[0].upper())
            else:
                 si=scale.index(self.tonic[0].upper() + self.tonic[1])   
        else:    
            si=scale.index(self.tonic)
    
        j=0
        s=[]
        while (si+j < len(scale)):
            #print ('j:',j,'si+j',si+j)
            s.append(scale[si+j])
            j += 1

        i=0
        while (i < si):
            s.append(scale[i])
            i += 1

        return s    

    def interval(self, intervals):
        scale=self.chromatic()
        #print ('scale=',scale)
        r=[]
        r.append(scale[0])
        j=0
        for i in intervals:
            if i == "M":
                j += 2
            elif i == "m":
                j += 1
            else:
                j += 3
            #print("j:",j)
            if j <= 11:
                r.append(scale[j])
        return r

File: solutions/go/scale-generator/1/scale_generator.go
//Package generates the musical scale
package scale
import "strings"

//Function generates teh musical scale
func Scale(tonic, interval string) []string {

	var snotes []string = []string{"A", "A#", "B","C", "C#", "D", "D#", "E", "F", "F#", "G", "G#"}
	var fnotes []string = []string{"A", "Bb", "B","C", "Db", "D", "Eb", "E", "F", "Gb", "G", "Ab"}
	var sharptonic []string = []string{"C","a","G","D","A","E","B","F#",
	                                   "e","b","f#","c#","g#","d#"}
	var scale []string

	if in(tonic,sharptonic) >= 0 {
		scale = snotes
	} else {
		scale = fnotes
	}
	if len(tonic) == 2 {
		tonic = strings.ToUpper(string(tonic[0]))+string(tonic[1])
	} else {
		tonic = strings.ToUpper(string(tonic[0]))
	}
	// find the position of the tonic
	if interval == "" {
		//chromatic scale
		return chromatic(tonic, scale)
	} else {
		//diatonic scale
		return diatonic(tonic,interval, scale)
	}
}

func in(item string, iarray []string) int {
var ii int
	for ii = 0; ii < len(iarray); ii++ {
		if item == iarray[ii] {
			return ii
		}
	}
	return -1
}

func chromatic(tonic string, scale []string) []string {
	var r []string
	var ti int = in(tonic, scale)
	for ni := ti; ni < len(scale); ni++ {
		r = append(r, scale[ni])
	}
	for ni := 0; ni < ti ; ni++ {
		r = append(r, scale[ni])
	}
	return r
}

func diatonic(tonic, interval string, scale []string) []string {
	var r []string
	var ti int = in(tonic, scale)

	r = append(r,scale[ti])

	for _, i := range interval {
		ti = ti + 1
		if ti > 11 {
			ti = 0
		}
		if string(i) == "M" {
			ti = ti + 1
			if ti > 11 {
				ti = 0
			}
		}
		if string(i) == "A" {
			ti = ti + 1
			if ti > 11 {
				ti = 0
			}
			ti = ti + 1
			if ti > 11 {
				ti = 0
			}
		}

		r = append(r,scale[ti])
	}
	r=r[:len(interval)]
	return r
}